Index: nsprpub/configure.in
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/configure.in,v
retrieving revision 1.199.2.14
diff -u -r1.199.2.14 configure.in
--- nsprpub/configure.in	27 Apr 2007 20:01:01 -0000	1.199.2.14
+++ nsprpub/configure.in	29 Dec 2007 15:13:38 -0000
@@ -629,6 +629,7 @@
         linux*)       OS_ARCH=Linux ;;
         solaris*)     OS_ARCH=SunOS OS_RELEASE=5 ;;
         mingw*)       OS_ARCH=WINNT ;;
+        riscos*)      OS_ARCH=RISCOS ;;
         darwin*)      OS_ARCH=Darwin ;;
     esac
 else
@@ -1800,7 +1801,7 @@
     MDCPUCFG_H=_riscos.cfg
     PR_MD_CSRCS=riscos.c
     DLL_SUFFIX=a
-    LD="/home/riscos/env/ro-ar cr"
+    MKSHLIB='$(AR) rc $@'
     ;;
 
 *-*-sco*)
Index: nsprpub/config/rules.mk
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/config/rules.mk,v
retrieving revision 3.61.2.1
diff -u -r3.61.2.1 rules.mk
--- nsprpub/config/rules.mk	22 Feb 2006 22:53:55 -0000	3.61.2.1
+++ nsprpub/config/rules.mk	29 Dec 2007 15:13:38 -0000
@@ -352,7 +352,11 @@
 	  fi; \
 	fi
 endif	# OpenVMS
+ifeq ($(OS_TARGET), RISCOS)
+	$(MKSHLIB) $(OBJS) $(RES)
+else
 	$(MKSHLIB) $(OBJS) $(RES) $(EXTRA_LIBS)
+endif
 endif   # OS2 vacpp
 endif	# WINNT
 endif	# AIX 4.1
Index: nsprpub/pr/src/malloc/prmem.c
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/pr/src/malloc/prmem.c,v
retrieving revision 3.19
diff -u -r3.19 prmem.c
--- nsprpub/pr/src/malloc/prmem.c	21 Jul 2005 18:22:53 -0000	3.19
+++ nsprpub/pr/src/malloc/prmem.c	29 Dec 2007 15:13:38 -0000
@@ -150,7 +150,7 @@
     return sym;
 }
 
-#elif defined(USE_MACH_DYLD)
+#elif defined(USE_MACH_DYLD) || defined(RISCOS)
 
 static void *
 pr_FindSymbolInProg(const char *name)
