? nsprpub/unallmakefiles
Index: nsprpub/configure
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/configure,v
retrieving revision 1.197.2.17
diff -u -r1.197.2.17 configure
--- nsprpub/configure	31 Oct 2007 18:07:38 -0000	1.197.2.17
+++ nsprpub/configure	18 Aug 2008 03:18:11 -0000
@@ -2753,6 +2753,7 @@
         linux*)       OS_ARCH=Linux ;;
         solaris*)     OS_ARCH=SunOS OS_RELEASE=5 ;;
         mingw*)       OS_ARCH=WINNT ;;
+        riscos*)      OS_ARCH=RISCOS ;;
         darwin*)      OS_ARCH=Darwin ;;
     esac
 else
@@ -2940,17 +2941,17 @@
     DSO_LDOPTS='-brtl -bnortllib -bM:SRE -bnoentry -bexpall -blibpath:/usr/lib:/lib'
     ac_safe=`echo "sys/atomic_op.h" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for sys/atomic_op.h""... $ac_c" 1>&6
-echo "configure:2944: checking for sys/atomic_op.h" >&5
+echo "configure:2945: checking for sys/atomic_op.h" >&5
 if eval "test \"`echo '$''{'ac_cv_header_$ac_safe'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 2949 "configure"
+#line 2950 "configure"
 #include "confdefs.h"
 #include <sys/atomic_op.h>
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
-{ (eval echo configure:2954: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
+{ (eval echo configure:2955: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
 ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
 if test -z "$ac_err"; then
   rm -rf conftest*
@@ -3107,7 +3108,7 @@
         _DEBUG_FLAGS='-gdwarf-2 -O0'
         MKSHLIB='$(CCC) $(DSO_LDOPTS) -o $@'
         echo $ac_n "checking for gethostbyaddr in -lbind""... $ac_c" 1>&6
-echo "configure:3111: checking for gethostbyaddr in -lbind" >&5
+echo "configure:3112: checking for gethostbyaddr in -lbind" >&5
 ac_lib_var=`echo bind'_'gethostbyaddr | sed 'y%./+-%__p_%'`
 if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -3115,7 +3116,7 @@
   ac_save_LIBS="$LIBS"
 LIBS="-lbind  $LIBS"
 cat > conftest.$ac_ext <<EOF
-#line 3119 "configure"
+#line 3120 "configure"
 #include "confdefs.h"
 /* Override any gcc2 internal prototype to avoid an error.  */
 /* We use char because int might match the return type of a gcc2
@@ -3126,7 +3127,7 @@
 gethostbyaddr()
 ; return 0; }
 EOF
-if { (eval echo configure:3130: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+if { (eval echo configure:3131: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   rm -rf conftest*
   eval "ac_cv_lib_$ac_lib_var=yes"
 else
@@ -4396,17 +4397,17 @@
         _OPTIMIZE_FLAGS="$_OPTIMIZE_FLAGS -Olimit 4000"
         ac_safe=`echo "machine/builtins.h" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for machine/builtins.h""... $ac_c" 1>&6
-echo "configure:4400: checking for machine/builtins.h" >&5
+echo "configure:4401: checking for machine/builtins.h" >&5
 if eval "test \"`echo '$''{'ac_cv_header_$ac_safe'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 4405 "configure"
+#line 4406 "configure"
 #include "confdefs.h"
 #include <machine/builtins.h>
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
-{ (eval echo configure:4410: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
+{ (eval echo configure:4411: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
 ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
 if test -z "$ac_err"; then
   rm -rf conftest*
@@ -4495,8 +4496,11 @@
     USE_PTHREADS=1
     MDCPUCFG_H=_riscos.cfg
     PR_MD_CSRCS=riscos.c
-    DLL_SUFFIX=a
-    LD="/home/riscos/env/ro-ar cr"
+    DSO_CFLAGS=-fPIC
+    DSO_LDOPTS='-shared -Wl,-soname -Wl,$(notdir $@)'
+    #DLL_SUFFIX=a
+    MKSHLIB='$(CC) $(DSO_LDOPTS) -o $@'
+    #MKSHLIB='$(AR) rc $@'
     ;;
 
 *-*-sco*)
@@ -4968,7 +4972,7 @@
     ;;
 *)
     echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
-echo "configure:4972: checking for dlopen in -ldl" >&5
+echo "configure:4976: checking for dlopen in -ldl" >&5
 ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
 if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -4976,7 +4980,7 @@
   ac_save_LIBS="$LIBS"
 LIBS="-ldl  $LIBS"
 cat > conftest.$ac_ext <<EOF
-#line 4980 "configure"
+#line 4984 "configure"
 #include "confdefs.h"
 /* Override any gcc2 internal prototype to avoid an error.  */
 /* We use char because int might match the return type of a gcc2
@@ -4987,7 +4991,7 @@
 dlopen()
 ; return 0; }
 EOF
-if { (eval echo configure:4991: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+if { (eval echo configure:4995: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   rm -rf conftest*
   eval "ac_cv_lib_$ac_lib_var=yes"
 else
@@ -5004,17 +5008,17 @@
   echo "$ac_t""yes" 1>&6
   ac_safe=`echo "dlfcn.h" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for dlfcn.h""... $ac_c" 1>&6
-echo "configure:5008: checking for dlfcn.h" >&5
+echo "configure:5012: checking for dlfcn.h" >&5
 if eval "test \"`echo '$''{'ac_cv_header_$ac_safe'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 5013 "configure"
+#line 5017 "configure"
 #include "confdefs.h"
 #include <dlfcn.h>
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
-{ (eval echo configure:5018: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
+{ (eval echo configure:5022: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
 ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
 if test -z "$ac_err"; then
   rm -rf conftest*
@@ -5047,13 +5051,13 @@
 
 if test $ac_cv_prog_gcc = yes; then
     echo $ac_n "checking whether ${CC-cc} needs -traditional""... $ac_c" 1>&6
-echo "configure:5051: checking whether ${CC-cc} needs -traditional" >&5
+echo "configure:5055: checking whether ${CC-cc} needs -traditional" >&5
 if eval "test \"`echo '$''{'ac_cv_prog_gcc_traditional'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
     ac_pattern="Autoconf.*'x'"
   cat > conftest.$ac_ext <<EOF
-#line 5057 "configure"
+#line 5061 "configure"
 #include "confdefs.h"
 #include <sgtty.h>
 Autoconf TIOCGETP
@@ -5071,7 +5075,7 @@
 
   if test $ac_cv_prog_gcc_traditional = no; then
     cat > conftest.$ac_ext <<EOF
-#line 5075 "configure"
+#line 5079 "configure"
 #include "confdefs.h"
 #include <termio.h>
 Autoconf TCGETA
@@ -5095,12 +5099,12 @@
 for ac_func in lchown strerror
 do
 echo $ac_n "checking for $ac_func""... $ac_c" 1>&6
-echo "configure:5099: checking for $ac_func" >&5
+echo "configure:5103: checking for $ac_func" >&5
 if eval "test \"`echo '$''{'ac_cv_func_$ac_func'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 5104 "configure"
+#line 5108 "configure"
 #include "confdefs.h"
 /* System header to define __stub macros and hopefully few prototypes,
     which can conflict with char $ac_func(); below.  */
@@ -5123,7 +5127,7 @@
 
 ; return 0; }
 EOF
-if { (eval echo configure:5127: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+if { (eval echo configure:5131: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   rm -rf conftest*
   eval "ac_cv_func_$ac_func=yes"
 else
@@ -5164,7 +5168,7 @@
 if test -z "$GNU_CC"; then
 
     echo $ac_n "checking for +Olit support""... $ac_c" 1>&6
-echo "configure:5168: checking for +Olit support" >&5
+echo "configure:5172: checking for +Olit support" >&5
 if eval "test \"`echo '$''{'ac_cv_hpux_usable_olit_option'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
@@ -5203,7 +5207,7 @@
 *)
     
 echo $ac_n "checking for pthread_create in -lpthreads""... $ac_c" 1>&6
-echo "configure:5207: checking for pthread_create in -lpthreads" >&5
+echo "configure:5211: checking for pthread_create in -lpthreads" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5225,7 +5229,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lpthread""... $ac_c" 1>&6
-echo "configure:5229: checking for pthread_create in -lpthread" >&5
+echo "configure:5233: checking for pthread_create in -lpthread" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5247,7 +5251,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lc_r""... $ac_c" 1>&6
-echo "configure:5251: checking for pthread_create in -lc_r" >&5
+echo "configure:5255: checking for pthread_create in -lc_r" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5269,7 +5273,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lc""... $ac_c" 1>&6
-echo "configure:5273: checking for pthread_create in -lc" >&5
+echo "configure:5277: checking for pthread_create in -lc" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5423,7 +5427,7 @@
       rm -f conftest*
    ac_cv_have_dash_pthread=no
    echo $ac_n "checking whether ${CC-cc} accepts -pthread""... $ac_c" 1>&6
-echo "configure:5427: checking whether ${CC-cc} accepts -pthread" >&5
+echo "configure:5431: checking whether ${CC-cc} accepts -pthread" >&5
    echo 'int main() { return 0; }' | cat > conftest.c
    ${CC-cc} -pthread -o conftest conftest.c > conftest.out 2>&1
    if test $? -eq 0; then
@@ -5446,7 +5450,7 @@
 			    ac_cv_have_dash_pthreads=no
     if test "$ac_cv_have_dash_pthread" = "no"; then
 	    echo $ac_n "checking whether ${CC-cc} accepts -pthreads""... $ac_c" 1>&6
-echo "configure:5450: checking whether ${CC-cc} accepts -pthreads" >&5
+echo "configure:5454: checking whether ${CC-cc} accepts -pthreads" >&5
     	echo 'int main() { return 0; }' | cat > conftest.c
 	    ${CC-cc} -pthreads -o conftest conftest.c > conftest.out 2>&1
     	if test $? -eq 0; then
@@ -5989,7 +5993,7 @@
 s%\]%\\&%g
 s%\$%$$%g
 EOF
-DEFS=`sed -f conftest.defs confdefs.h | tr '\012' ' ' | tr '\015' ' '`
+DEFS=`sed -f conftest.defs confdefs.h | tr '\012' ' '`
 rm -f conftest.defs
 
 
Index: nsprpub/configure.in
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/configure.in,v
retrieving revision 1.199.2.17
diff -u -r1.199.2.17 configure.in
--- nsprpub/configure.in	31 Oct 2007 18:07:38 -0000	1.199.2.17
+++ nsprpub/configure.in	18 Aug 2008 03:18:11 -0000
@@ -629,6 +629,7 @@
         linux*)       OS_ARCH=Linux ;;
         solaris*)     OS_ARCH=SunOS OS_RELEASE=5 ;;
         mingw*)       OS_ARCH=WINNT ;;
+        riscos*)      OS_ARCH=RISCOS ;;
         darwin*)      OS_ARCH=Darwin ;;
     esac
 else
@@ -1814,8 +1815,11 @@
     USE_PTHREADS=1
     MDCPUCFG_H=_riscos.cfg
     PR_MD_CSRCS=riscos.c
-    DLL_SUFFIX=a
-    LD="/home/riscos/env/ro-ar cr"
+    DSO_CFLAGS=-fPIC
+    DSO_LDOPTS='-shared -Wl,-soname -Wl,$(notdir $@)'
+    #DLL_SUFFIX=a
+    MKSHLIB='$(CC) $(DSO_LDOPTS) -o $@'
+    #MKSHLIB='$(AR) rc $@'
     ;;
 
 *-*-sco*)
Index: nsprpub/config/rules.mk
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/config/rules.mk,v
retrieving revision 3.61.2.1
diff -u -r3.61.2.1 rules.mk
--- nsprpub/config/rules.mk	22 Feb 2006 22:53:55 -0000	3.61.2.1
+++ nsprpub/config/rules.mk	18 Aug 2008 03:18:11 -0000
@@ -352,7 +352,11 @@
 	  fi; \
 	fi
 endif	# OpenVMS
+ifeq ($(OS_TARGET), RISCOS)
+	$(MKSHLIB) $(OBJS) $(RES)
+else
 	$(MKSHLIB) $(OBJS) $(RES) $(EXTRA_LIBS)
+endif
 endif   # OS2 vacpp
 endif	# WINNT
 endif	# AIX 4.1
Index: nsprpub/pr/include/md/_riscos.h
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/pr/include/md/_riscos.h,v
retrieving revision 3.2
diff -u -r3.2 _riscos.h
--- nsprpub/pr/include/md/_riscos.h	21 Jul 2005 18:22:53 -0000	3.2
+++ nsprpub/pr/include/md/_riscos.h	18 Aug 2008 03:18:11 -0000
@@ -44,7 +44,7 @@
 #define PR_LINKER_ARCH		"riscos"
 #define _PR_SI_SYSNAME		"RISCOS"
 #define _PR_SI_ARCHITECTURE	"arm"
-#define PR_DLL_SUFFIX		".a"
+#define PR_DLL_SUFFIX		".so"
 
 #define _PR_POLL_AVAILABLE
 #define _PR_USE_POLL
@@ -60,8 +60,8 @@
 
 
 #undef  HAVE_STACK_GROWING_UP
-#undef  HAVE_DLL
-#undef  USE_DLFCN
+#define HAVE_DLL
+#define USE_DLFCN
 #define NEED_STRFTIME_LOCK
 #define NEED_TIME_R
 #define PT_NO_SIGTIMEDWAIT
Index: nsprpub/pr/src/linking/prlink.c
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/pr/src/linking/prlink.c,v
retrieving revision 3.81.2.5
diff -u -r3.81.2.5 prlink.c
--- nsprpub/pr/src/linking/prlink.c	6 Oct 2006 23:36:22 -0000	3.81.2.5
+++ nsprpub/pr/src/linking/prlink.c	18 Aug 2008 03:18:12 -0000
@@ -964,6 +964,7 @@
         dl_flags |= RTLD_LOCAL;
     }
     h = dlopen(name, dl_flags);
+    if (!h) perror("fail");
 #elif defined(USE_HPSHL)
     int shl_flags = 0;
     shl_t h;
Index: nsprpub/pr/src/misc/prenv.c
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/pr/src/misc/prenv.c,v
retrieving revision 3.11
diff -u -r3.11 prenv.c
--- nsprpub/pr/src/misc/prenv.c	25 Apr 2004 15:01:01 -0000	3.11
+++ nsprpub/pr/src/misc/prenv.c	18 Aug 2008 03:18:12 -0000
@@ -79,7 +79,16 @@
     if (!_pr_initialized) _PR_ImplicitInitialization();
 
     _PR_LOCK_ENV();
+#ifdef __riscos__
+    {
+      char *roenv = alloca(strlen(var) + 9);
+      strcpy(roenv, "Firefox$");
+      strcpy(roenv + 8, var);
+      ev = _PR_MD_GET_ENV(roenv);
+    }
+#else
     ev = _PR_MD_GET_ENV(var);
+#endif
     _PR_UNLOCK_ENV();
     return ev;
 }
@@ -93,7 +102,16 @@
     if ( !strchr(string, '=')) return(PR_FAILURE);
 
     _PR_LOCK_ENV();
+#ifdef __riscos__
+    {
+      char *roenv = alloca(strlen(string) + 9);
+      strcpy(roenv, "Firefox$");
+      strcpy(roenv + 8, string);
+      result = _PR_MD_PUT_ENV(roenv);
+    }
+#else
     result = _PR_MD_PUT_ENV(string);
+#endif
     _PR_UNLOCK_ENV();
     return (result)? PR_FAILURE : PR_SUCCESS;
 }
