Index: security/coreconf/nsinstall/nsinstall.c
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/nsinstall/nsinstall.c,v
retrieving revision 1.8
diff -u -r1.8 nsinstall.c
--- security/coreconf/nsinstall/nsinstall.c	25 Apr 2004 15:02:18 -0000	1.8
+++ security/coreconf/nsinstall/nsinstall.c	26 Aug 2008 03:14:46 -0000
@@ -58,7 +58,7 @@
 
 #define HAVE_LCHOWN
 
-#if defined(AIX) || defined(BSDI) || defined(HPUX) || defined(LINUX) || defined(SUNOS4) || defined(SCO) || defined(UNIXWARE) || defined(VMS) || defined(NTO) || defined(DARWIN) || defined(BEOS)
+#if defined(AIX) || defined(BSDI) || defined(HPUX) || defined(LINUX) || defined(SUNOS4) || defined(SCO) || defined(UNIXWARE) || defined(VMS) || defined(NTO) || defined(DARWIN) || defined(BEOS) || defined(__riscos__)
 #undef HAVE_LCHOWN
 #endif
 
Index: security/nss-fips/lib/freebl/unix_rand.c
===================================================================
RCS file: /cvsroot/mozilla/security/nss/lib/freebl/unix_rand.c,v
retrieving revision 1.17.10.5
diff -u -r1.17.10.5 unix_rand.c
--- security/nss-fips/lib/freebl/unix_rand.c	12 Oct 2006 22:50:48 -0000	1.17.10.5
+++ security/nss-fips/lib/freebl/unix_rand.c	26 Aug 2008 03:14:47 -0000
@@ -365,7 +365,7 @@
 
 #endif /* Alpha */
 
-#if defined(_IBMR2)
+#if defined(_IBMR2) || defined(__riscos__)
 static size_t
 GetHighResClock(void *buf, size_t maxbytes)
 {
