diff -r 11bdfa7b7c58 nsprpub/configure
--- a/nsprpub/configure	Thu Aug 06 15:43:07 2009 +0200
+++ b/nsprpub/configure	Thu Aug 06 11:58:16 2009 -0700
@@ -568,7 +568,7 @@
 ac_link='${CC-cc} -o conftest${ac_exeext} $CFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS 1>&5'
 cross_compiling=$ac_cv_prog_cc_cross
 
-ac_exeext=
+ac_exeext=,e1f;
 ac_objext=o
 if (echo "testing\c"; echo 1,2,3) | grep c >/dev/null; then
   # Stardent Vistra SVR4 grep lacks -e, says ghazi@caip.rutgers.edu.
@@ -4604,8 +4604,11 @@
     USE_PTHREADS=1
     MDCPUCFG_H=_riscos.cfg
     PR_MD_CSRCS=riscos.c
-    DLL_SUFFIX=a
-    LD="/home/riscos/env/ro-ar cr"
+    DSO_CFLAGS=-fPIC
+    DSO_LDOPTS='-shared -Wl,-soname -Wl,$(notdir $@)'
+    #DLL_SUFFIX=a
+    MKSHLIB='$(CC) $(DSO_LDOPTS) -o $@'
+    #MKSHLIB='$(AR) rc $@'
     ;;
 
 *-*-sco*)
@@ -5164,7 +5167,7 @@
     ;;
 *)
     echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
-echo "configure:5168: checking for dlopen in -ldl" >&5
+echo "configure:5171: checking for dlopen in -ldl" >&5
 ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
 if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -5172,7 +5175,7 @@
   ac_save_LIBS="$LIBS"
 LIBS="-ldl  $LIBS"
 cat > conftest.$ac_ext <<EOF
-#line 5176 "configure"
+#line 5179 "configure"
 #include "confdefs.h"
 /* Override any gcc2 internal prototype to avoid an error.  */
 /* We use char because int might match the return type of a gcc2
@@ -5183,7 +5186,7 @@
 dlopen()
 ; return 0; }
 EOF
-if { (eval echo configure:5187: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+if { (eval echo configure:5190: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   rm -rf conftest*
   eval "ac_cv_lib_$ac_lib_var=yes"
 else
@@ -5200,17 +5203,17 @@
   echo "$ac_t""yes" 1>&6
   ac_safe=`echo "dlfcn.h" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for dlfcn.h""... $ac_c" 1>&6
-echo "configure:5204: checking for dlfcn.h" >&5
+echo "configure:5207: checking for dlfcn.h" >&5
 if eval "test \"`echo '$''{'ac_cv_header_$ac_safe'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 5209 "configure"
+#line 5212 "configure"
 #include "confdefs.h"
 #include <dlfcn.h>
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
-{ (eval echo configure:5214: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
+{ (eval echo configure:5217: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
 ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
 if test -z "$ac_err"; then
   rm -rf conftest*
@@ -5243,13 +5246,13 @@
 
 if test $ac_cv_prog_gcc = yes; then
     echo $ac_n "checking whether ${CC-cc} needs -traditional""... $ac_c" 1>&6
-echo "configure:5247: checking whether ${CC-cc} needs -traditional" >&5
+echo "configure:5250: checking whether ${CC-cc} needs -traditional" >&5
 if eval "test \"`echo '$''{'ac_cv_prog_gcc_traditional'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
     ac_pattern="Autoconf.*'x'"
   cat > conftest.$ac_ext <<EOF
-#line 5253 "configure"
+#line 5256 "configure"
 #include "confdefs.h"
 #include <sgtty.h>
 Autoconf TIOCGETP
@@ -5267,7 +5270,7 @@
 
   if test $ac_cv_prog_gcc_traditional = no; then
     cat > conftest.$ac_ext <<EOF
-#line 5271 "configure"
+#line 5274 "configure"
 #include "confdefs.h"
 #include <termio.h>
 Autoconf TCGETA
@@ -5291,12 +5294,12 @@
 for ac_func in lchown strerror
 do
 echo $ac_n "checking for $ac_func""... $ac_c" 1>&6
-echo "configure:5295: checking for $ac_func" >&5
+echo "configure:5298: checking for $ac_func" >&5
 if eval "test \"`echo '$''{'ac_cv_func_$ac_func'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   cat > conftest.$ac_ext <<EOF
-#line 5300 "configure"
+#line 5303 "configure"
 #include "confdefs.h"
 /* System header to define __stub macros and hopefully few prototypes,
     which can conflict with char $ac_func(); below.  */
@@ -5319,7 +5322,7 @@
 
 ; return 0; }
 EOF
-if { (eval echo configure:5323: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+if { (eval echo configure:5326: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   rm -rf conftest*
   eval "ac_cv_func_$ac_func=yes"
 else
@@ -5360,14 +5363,14 @@
 if test -z "$GNU_CC"; then
 
     echo $ac_n "checking for +Olit support""... $ac_c" 1>&6
-echo "configure:5364: checking for +Olit support" >&5
+echo "configure:5367: checking for +Olit support" >&5
 if eval "test \"`echo '$''{'ac_cv_hpux_usable_olit_option'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
                   ac_cv_hpux_usable_olit_option=no
         rm -f conftest*
         echo 'int main() { return 0; }' | cat > conftest.c
-        ${CC-cc} ${CFLAGS} +Olit=all -o conftest conftest.c > conftest.out 2>&1
+        ${CC-cc} ${CFLAGS} +Olit=all -o conftest${ac_exeext} conftest.c > conftest.out 2>&1
         if test $? -eq 0; then
             if test -z "`egrep -i '(unrecognize|unknown)' conftest.out`"; then
                 ac_cv_hpux_usable_olit_option=yes
@@ -5399,7 +5402,7 @@
 *)
     
 echo $ac_n "checking for pthread_create in -lpthreads""... $ac_c" 1>&6
-echo "configure:5403: checking for pthread_create in -lpthreads" >&5
+echo "configure:5406: checking for pthread_create in -lpthreads" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5421,7 +5424,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lpthread""... $ac_c" 1>&6
-echo "configure:5425: checking for pthread_create in -lpthread" >&5
+echo "configure:5428: checking for pthread_create in -lpthread" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5443,7 +5446,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lc_r""... $ac_c" 1>&6
-echo "configure:5447: checking for pthread_create in -lc_r" >&5
+echo "configure:5450: checking for pthread_create in -lc_r" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5465,7 +5468,7 @@
         echo "$ac_t""no" 1>&6
         
 echo $ac_n "checking for pthread_create in -lc""... $ac_c" 1>&6
-echo "configure:5469: checking for pthread_create in -lc" >&5
+echo "configure:5472: checking for pthread_create in -lc" >&5
 echo "
     #include <pthread.h> 
     void *foo(void *v) { return v; } 
@@ -5597,9 +5600,9 @@
       rm -f conftest*
    ac_cv_have_dash_pthread=no
    echo $ac_n "checking whether ${CC-cc} accepts -pthread""... $ac_c" 1>&6
-echo "configure:5601: checking whether ${CC-cc} accepts -pthread" >&5
+echo "configure:5604: checking whether ${CC-cc} accepts -pthread" >&5
    echo 'int main() { return 0; }' | cat > conftest.c
-   ${CC-cc} -pthread -o conftest conftest.c > conftest.out 2>&1
+   ${CC-cc} -pthread -o conftest${ac_exeext} conftest.c > conftest.out 2>&1
    if test $? -eq 0; then
 	if test -z "`egrep -i '(unrecognize|unknown)' conftest.out | grep pthread`" && test -z "`egrep -i '(error|incorrect)' conftest.out`" ; then
 	    ac_cv_have_dash_pthread=yes
@@ -5620,9 +5623,9 @@
 			    ac_cv_have_dash_pthreads=no
     if test "$ac_cv_have_dash_pthread" = "no"; then
 	    echo $ac_n "checking whether ${CC-cc} accepts -pthreads""... $ac_c" 1>&6
-echo "configure:5624: checking whether ${CC-cc} accepts -pthreads" >&5
+echo "configure:5627: checking whether ${CC-cc} accepts -pthreads" >&5
     	echo 'int main() { return 0; }' | cat > conftest.c
-	    ${CC-cc} -pthreads -o conftest conftest.c > conftest.out 2>&1
+	    ${CC-cc} -pthreads -o conftest${ac_exeext} conftest.c > conftest.out 2>&1
     	if test $? -eq 0; then
 	    	if test -z "`egrep -i '(unrecognize|unknown)' conftest.out | grep pthreads`" && test -z "`egrep -i '(error|incorrect)' conftest.out`" ; then
 			    ac_cv_have_dash_pthreads=yes
@@ -6154,7 +6157,7 @@
 s%\]%\\&%g
 s%\$%$$%g
 EOF
-DEFS=`sed -f conftest.defs confdefs.h | tr '\012' ' ' | tr '\015' ' '` # Manually modified for MKS support.
+DEFS=`sed -f conftest.defs confdefs.h | tr '\012' ' '`
 rm -f conftest.defs
 
 
diff -r 11bdfa7b7c58 nsprpub/configure.in
--- a/nsprpub/configure.in	Thu Aug 06 15:43:07 2009 +0200
+++ b/nsprpub/configure.in	Thu Aug 06 11:58:16 2009 -0700
@@ -1880,8 +1880,11 @@
     USE_PTHREADS=1
     MDCPUCFG_H=_riscos.cfg
     PR_MD_CSRCS=riscos.c
-    DLL_SUFFIX=a
-    LD="/home/riscos/env/ro-ar cr"
+    DSO_CFLAGS=-fPIC
+    DSO_LDOPTS='-shared -Wl,-soname -Wl,$(notdir $@)'
+    #DLL_SUFFIX=a
+    MKSHLIB='$(CC) $(DSO_LDOPTS) -o $@'
+    #MKSHLIB='$(AR) rc $@'
     ;;
 
 *-*-sco*)
diff -r 11bdfa7b7c58 nsprpub/pr/include/md/_riscos.h
--- a/nsprpub/pr/include/md/_riscos.h	Thu Aug 06 15:43:07 2009 +0200
+++ b/nsprpub/pr/include/md/_riscos.h	Thu Aug 06 11:58:16 2009 -0700
@@ -44,7 +44,7 @@
 #define PR_LINKER_ARCH		"riscos"
 #define _PR_SI_SYSNAME		"RISCOS"
 #define _PR_SI_ARCHITECTURE	"arm"
-#define PR_DLL_SUFFIX		".a"
+#define PR_DLL_SUFFIX		".so"
 
 #define _PR_POLL_AVAILABLE
 #define _PR_USE_POLL
@@ -60,8 +60,8 @@
 
 
 #undef  HAVE_STACK_GROWING_UP
-#undef  HAVE_DLL
-#undef  USE_DLFCN
+#define HAVE_DLL
+#define USE_DLFCN
 #define NEED_STRFTIME_LOCK
 #define NEED_TIME_R
 #define PT_NO_SIGTIMEDWAIT
diff -r 11bdfa7b7c58 nsprpub/pr/src/misc/prenv.c
--- a/nsprpub/pr/src/misc/prenv.c	Thu Aug 06 15:43:07 2009 +0200
+++ b/nsprpub/pr/src/misc/prenv.c	Thu Aug 06 11:58:16 2009 -0700
@@ -79,7 +79,16 @@
     if (!_pr_initialized) _PR_ImplicitInitialization();
 
     _PR_LOCK_ENV();
+#ifdef __riscos__
+    {
+      char *roenv = alloca(strlen(var) + 9);
+      strcpy(roenv, "Firefox$");
+      strcpy(roenv + 8, var);
+      ev = _PR_MD_GET_ENV(roenv);
+    }
+#else
     ev = _PR_MD_GET_ENV(var);
+#endif
     _PR_UNLOCK_ENV();
     return ev;
 }
@@ -93,7 +102,16 @@
     if ( !strchr(string, '=')) return(PR_FAILURE);
 
     _PR_LOCK_ENV();
+#ifdef __riscos__
+    {
+      char *roenv = alloca(strlen(string) + 9);
+      strcpy(roenv, "Firefox$");
+      strcpy(roenv + 8, string);
+      result = _PR_MD_PUT_ENV(roenv);
+    }
+#else
     result = _PR_MD_PUT_ENV(string);
+#endif
     _PR_UNLOCK_ENV();
     return (result)? PR_FAILURE : PR_SUCCESS;
 }
diff -r 11bdfa7b7c58 nsprpub/pr/tests/Makefile.in
--- a/nsprpub/pr/tests/Makefile.in	Thu Aug 06 15:43:07 2009 +0200
+++ b/nsprpub/pr/tests/Makefile.in	Thu Aug 06 11:58:16 2009 -0700
@@ -456,6 +456,11 @@
 endif
 endif
 
+ifeq ($(OS_ARCH),RISCOS)
+EXTRA_LIBS = -ldl
+endif
+
+
 ifeq ($(USE_PTHREADS),1)
 LIBPTHREAD = -lpthread
 ifeq ($(OS_ARCH),AIX)
