--- ../../../../gcc3.4.6/gcc-3.4.6/gcc/config/arm/arm-protos.h	2005-10-01 15:31:38.000000000 +0200
+++ ./config/arm/arm-protos.h	2006-04-15 15:21:14.000000000 +0200
@@ -152,7 +152,7 @@
 #endif
 
 #if defined AOF_ASSEMBLER 
-extern rtx aof_pic_entry (rtx);
+extern rtx aof_pic_entry (rtx, int);
 extern char *aof_text_section (void);
 extern char *aof_data_section (void);
 extern void aof_add_import (const char *);
