--- ../../../gcc-3.4-20050719/gcc/c-opts.c	2005-03-09 02:00:56.000000000 +0100
+++ ./c-opts.c	2005-03-21 12:56:18.000000000 +0100
@@ -151,6 +151,7 @@
       break;
 
     case OPT_I:
+    case OPT_icrossdirafter: /* NAB */
     case OPT_idirafter:
     case OPT_isysroot:
     case OPT_isystem:
@@ -937,6 +938,26 @@
       add_path (xstrdup (arg), AFTER, 0);
       break;
 
+      /* NAB++ */
+    case OPT_icrossdirafter:
+      {
+	char *path;
+	const char *prefix;
+	const char *suffix = arg;
+	size_t prefix_len, suffix_len;
+
+	suffix_len = strlen (suffix);
+	prefix     = cpp_CROSS_INCLUDE_DIR;
+	prefix_len = cpp_CROSS_INCLUDE_DIR_len;
+
+	path = xmalloc (prefix_len + suffix_len + 1);
+	strcpy (path, prefix);
+	strcat (path, suffix);
+	add_path (path, AFTER, 0);
+      }
+      break;
+      /* NAB-- */
+
     case OPT_imacros:
     case OPT_include:
       defer_opt (code, arg);
