--- ../../../gcc-3.4-20050719/gcc/c.opt	2005-03-19 21:30:38.000000000 +0100
+++ ./c.opt	2005-08-10 02:38:48.000000000 +0200
@@ -688,6 +688,10 @@
 ObjC ObjC++
 Dump declarations to a .decl file
 
+icrossdirafter
+C ObjC C++ ObjC++ Joined Separate
+-icrossdirafter <dir>	Add <dir> to the end of the system include path
+
 idirafter
 C ObjC C++ ObjC++ Joined Separate
 -idirafter <dir>	Add <dir> to the end of the system include path
