--- /home/nick/riscos/masters/gcc-3.4/gcc/c.opt	2004-02-18 00:09:03.000000000 +0000
+++ c.opt	2004-05-31 16:10:22.000000000 +0100
@@ -688,6 +688,10 @@ gen-decls
 ObjC ObjC++
 Dump declarations to a .decl file
 
+icrossdirafter
+C ObjC C++ ObjC++ Joined Separate
+-icrossdirafter <dir>	Add <dir> to the end of the system include path
+
 idirafter
 C ObjC C++ ObjC++ Joined Separate
 -idirafter <dir>	Add <dir> to the end of the system include path
