Index: cpplib.c
===================================================================
RCS file: /usr/local/cvsroot/gccsdk/gcc/gcc/cpplib.c,v
retrieving revision 1.8
diff -u -r1.8 cpplib.c
--- cpplib.c	31 May 2004 15:10:24 -0000	1.8
+++ cpplib.c	25 Sep 2004 15:43:48 -0000
@@ -636,6 +636,11 @@
 {
   char *fname;
   const cpp_token *header;
+#if defined(CROSS_COMPILE)
+  char unixed[256]; /* Temp buffer to translate Unix filenames to RISCOS */
+  int len;
+  extern char *riscos_to_unix (char *, char *);
+#endif
 
   /* Allow macro expansion.  */
   header = get_token_no_padding (pfile);
@@ -644,6 +649,14 @@
       fname = xmalloc (header->val.str.len - 1);
       memcpy (fname, header->val.str.text + 1, header->val.str.len - 2);
       fname[header->val.str.len - 2] = '\0';
+#if defined(CROSS_COMPILE)
+      /* Convert the #include filename to a Unix version.  */
+      riscos_to_unix (fname, unixed);
+      free (fname);
+      len = strlen (unixed) + 1;
+      fname = xmalloc (len);
+      memcpy (fname, unixed, len);
+#endif
       *pangle_brackets = header->type == CPP_HEADER_NAME;
     }
   else if (header->type == CPP_LESS)
