--- ../../../gcc-3.4-20050719/gcc/diagnostic.c	2003-09-22 07:09:12.000000000 +0200
+++ ./diagnostic.c	2004-06-01 20:33:52.000000000 +0200
@@ -355,7 +355,24 @@
     {
       (*diagnostic_starter (context)) (context, diagnostic);
       pp_format_text (context->printer, &diagnostic->message);
+
+#ifdef ERROR_THROWBACK
+      {
+	const char *s;
+
+	/* pp_base_format_verbatim (context->printer, &diagnostic->message); */
+	s = pp_formatted_text (context->printer);
+	/* printf ("throwback = '%s'\n", s); */
+	ERROR_THROWBACK (diagnostic->location.file,
+			 diagnostic->location.line,
+			 (diagnostic->kind == DK_WARNING) ? "warning" : NULL,
+			 s);
+	/* free (s); */
+      }
+#endif
+
       (*diagnostic_finalizer (context)) (context, diagnostic);
+
       pp_flush (context->printer);
       diagnostic_action_after_output (context, diagnostic);
     }
