--- ../../../gcc-3.4-20050719/gcc/gcc.h	2003-07-06 08:15:35.000000000 +0200
+++ ./gcc.h	2004-05-31 17:10:31.000000000 +0200
@@ -50,6 +50,8 @@
   || !strcmp (STR, "iwithprefix") || !strcmp (STR, "iwithprefixbefore") \
   || !strcmp (STR, "isystem") || !strcmp (STR, "-param") \
   || !strcmp (STR, "specs") \
+  || !strcmp (STR, "icrossdirafter") /* NAB++ */ \
+  || !strcmp (STR, "icrossdirsystem") /* NAB++ */ \
   || !strcmp (STR, "MF") || !strcmp (STR, "MT") || !strcmp (STR, "MQ"))
 
 
