--- ../../../gcc-3.4-20050719/gcc/loop-unswitch.c	2003-12-30 11:40:54.000000000 +0100
+++ ./loop-unswitch.c	2005-07-03 13:06:11.000000000 +0200
@@ -286,6 +286,13 @@
 	}
     } while (repeat);
 
+  if (BASIC_BLOCK (BLOCK_NUM (PREV_INSN (split_before))) != bbs[i])
+    {
+      if (rtl_dump_file)
+	fprintf (rtl_dump_file, ";; Not unswitching, cond in different bb\n");
+      return;
+    }
+      
   /* We found the condition we can unswitch on.  */
   conds = alloc_EXPR_LIST (0, cond, cond_checked);
   if (rcond)
