Index: configure.in
===================================================================
RCS file: /cvsroot/mozilla/configure.in,v
retrieving revision 1.1503.2.15.2.8
diff -u -r1.1503.2.15.2.8 configure.in
--- configure.in	21 Apr 2006 20:50:21 -0000	1.1503.2.15.2.8
+++ configure.in	22 May 2006 00:39:31 -0000
@@ -773,7 +773,6 @@
 CYGWIN_WRAPPER=
 WIN_TOP_SRC=
 MOZ_USER_DIR=".mozilla"
-HOST_AR='$(AR)'
 HOST_AR_FLAGS='$(AR_FLAGS)'
 
 MOZ_JPEG_CFLAGS=
@@ -1277,6 +1276,20 @@
     HOST_OPTIMIZE_FLAGS="${HOST_OPTIMIZE_FLAGS=-O2}"
     ;;
 
+*riscos*)
+    HOST_CFLAGS="$HOST_CFLAGS -DXP_UNIX"
+    HOST_OPTIMIZE_FLAGS="${HOST_OPTIMIZE_FLAGS=-O2}"
+    DSO_PIC_CFLAGS=
+    DSO_LDOPTS=
+    DLL_SUFFIX=.a
+    BIN_SUFFIX=,ff8
+    MKSHLIB="$GCCSDK_INSTALL_ENV/ro-ar rc \$@"
+    MKCSHLIB="$GCCSDK_INSTALL_ENV/ro-ar rc \$@"
+    MKSHLIB_FORCE_ALL=''
+    MKSHLIB_UNFORCE_ALL=''
+    MOZ_OPTIMIZE_FLAGS=-O2
+    ;;
+
 *)
     HOST_CFLAGS="$HOST_CFLAGS -DXP_UNIX"
     HOST_OPTIMIZE_FLAGS="${HOST_OPTIMIZE_FLAGS=-O2}"
Index: browser/app/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/browser/app/Makefile.in,v
retrieving revision 1.85.2.5
diff -u -r1.85.2.5 Makefile.in
--- browser/app/Makefile.in	7 Sep 2005 01:51:58 -0000	1.85.2.5
+++ browser/app/Makefile.in	22 May 2006 00:39:32 -0000
@@ -72,7 +72,7 @@
 LOCAL_INCLUDES += -I$(topsrcdir)/toolkit/xre
 
 ifdef BUILD_STATIC_LIBS
-STATIC_COMPONENTS_LINKER_PATH = -L$(DIST)/lib/components
+STATIC_COMPONENTS_LINKER_PATH = -L$(DIST)/lib/components -lSM -lICE -ldb-4.4 -lfreebl
 endif
 
 ifneq (,$(filter mac cocoa,$(MOZ_WIDGET_TOOLKIT)))
Index: config/rules.mk
===================================================================
RCS file: /cvsroot/mozilla/config/rules.mk,v
retrieving revision 3.487.2.1.2.2
diff -u -r3.487.2.1.2.2 rules.mk
--- config/rules.mk	28 Apr 2006 19:25:51 -0000	3.487.2.1.2.2
+++ config/rules.mk	22 May 2006 00:39:35 -0000
@@ -40,6 +40,8 @@
 topsrcdir		= $(DEPTH)
 endif
 
+DISABLE_DIST_GRE=1
+
 ifndef MOZILLA_DIR
 MOZILLA_DIR = $(topsrcdir)
 endif
Index: dbm/include/mcom_db.h
===================================================================
RCS file: /cvsroot/mozilla/dbm/include/mcom_db.h,v
retrieving revision 3.41
diff -u -r3.41 mcom_db.h
--- dbm/include/mcom_db.h	29 Apr 2005 13:33:34 -0000	3.41
+++ dbm/include/mcom_db.h	22 May 2006 00:39:37 -0000
@@ -329,6 +329,8 @@
 #define	DB_TXN		    0x8000	/* Do transactions. */
 #endif
 
+#define uint unsigned int
+
 /* Access method description structure. */
 typedef struct __db {
 	DBTYPE type;			/* Underlying db type. */
Index: dbm/src/mktemp.c
===================================================================
RCS file: /cvsroot/mozilla/dbm/src/mktemp.c,v
retrieving revision 3.5
diff -u -r3.5 mktemp.c
--- dbm/src/mktemp.c	20 Jan 2003 23:13:37 -0000	3.5
+++ dbm/src/mktemp.c	22 May 2006 00:39:37 -0000
@@ -99,7 +99,7 @@
 _gettemp(char *path, register int *doopen, int extraFlags)
 {    
 #if !defined(_WINDOWS) || defined(_WIN32)
-	extern int errno;                    
+	//extern int errno;
 #endif
 	register char *start, *trv;
 	struct stat sbuf;
Index: directory/c-sdk/ldap/libraries/libprldap/ldappr-error.c
===================================================================
RCS file: /cvsroot/mozilla/directory/c-sdk/ldap/libraries/libprldap/ldappr-error.c,v
retrieving revision 5.0.2.9
diff -u -r5.0.2.9 ldappr-error.c
--- directory/c-sdk/ldap/libraries/libprldap/ldappr-error.c	25 Mar 2003 13:57:38 -0000	5.0.2.9
+++ directory/c-sdk/ldap/libraries/libprldap/ldappr-error.c	22 May 2006 00:39:42 -0000
@@ -208,17 +208,17 @@
 #define ETXTBSY         -1
 #endif
 
-#if defined(BSDI) || defined(OPENBSD) || defined (NETBSD)
+#if defined(BSDI) || defined(OPENBSD) || defined (NETBSD) || defined (__riscos__)
 #define ENOTSUP		-1
 #endif
 
-#if defined(OSF1) || defined(BSDI) || defined(VMS) || defined(OPENBSD)
+#if defined(OSF1) || defined(BSDI) || defined(VMS) || defined(OPENBSD) || defined (__riscos__)
 #define EOVERFLOW       -1
 #endif
 
 #if defined(__hpux) || defined(_AIX) || defined(OSF1) || defined(DARWIN) || \
   defined(BEOS) || defined(FREEBSD) || defined(BSDI) || defined(VMS) || \
-  defined(OPENBSD) || defined(NETBSD)
+  defined(OPENBSD) || defined(NETBSD) || defined (__riscos__)
 #define EDEADLOCK       -1
 #endif
 
Index: gfx/src/x11shared/nsFontDebug.h
===================================================================
RCS file: /cvsroot/mozilla/gfx/src/x11shared/nsFontDebug.h,v
retrieving revision 1.3
diff -u -r1.3 nsFontDebug.h
--- gfx/src/x11shared/nsFontDebug.h	20 Aug 2004 09:11:25 -0000	1.3
+++ gfx/src/x11shared/nsFontDebug.h	22 May 2006 00:40:03 -0000
@@ -52,7 +52,7 @@
 #define NS_FONT_DEBUG_FREETYPE_GRAPHICS 0x800
 
 #undef NS_FONT_DEBUG
-#define NS_FONT_DEBUG 1
+//#define NS_FONT_DEBUG 1
 #ifdef NS_FONT_DEBUG
 
 # define DEBUG_PRINTF(x) \
Index: gfx/src/x11shared/nsX11AlphaBlend.cpp
===================================================================
RCS file: /cvsroot/mozilla/gfx/src/x11shared/nsX11AlphaBlend.cpp,v
retrieving revision 1.4
diff -u -r1.4 nsX11AlphaBlend.cpp
--- gfx/src/x11shared/nsX11AlphaBlend.cpp	17 Apr 2004 21:52:34 -0000	1.4
+++ gfx/src/x11shared/nsX11AlphaBlend.cpp	22 May 2006 00:40:04 -0000
@@ -48,7 +48,7 @@
 #include "nsAntiAliasedGlyph.h"
 #include "nsX11AlphaBlend.h"
 
-#define ENABLE_X11ALPHA_BLEND_PRINTF 1
+//#define ENABLE_X11ALPHA_BLEND_PRINTF 1
 
 #if ENABLE_X11ALPHA_BLEND_PRINTF
 static PRUint32 gX11AlphaBlendDebug;
Index: js/src/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/js/src/Makefile.in,v
retrieving revision 3.95.12.1
diff -u -r3.95.12.1 Makefile.in
--- js/src/Makefile.in	23 Feb 2006 19:30:43 -0000	3.95.12.1
+++ js/src/Makefile.in	22 May 2006 00:40:08 -0000
@@ -307,7 +307,7 @@
 
 # On OS/2 & win32 we are already linking against fdlibm, so don't bother
 # creating jsmathtemp
-ifeq (,$(filter OS2 WINNT OpenVMS,$(OS_ARCH)))
+ifeq (,$(filter OS2 WINNT OpenVMS riscos,$(OS_ARCH)))
 # special rule for jsmath.o since we want to incrementally link
 # against fdlibm to pull in only what is needed
 # Do this in a single step to avoid dependency problems
@@ -322,7 +322,8 @@
 	$(LD) $(DASH_R) -o $@ $(JSMATH_PRELINK) $(FDLIBM_LIBRARY)
 endif
 else
-	$(LD) $(DASH_R) -o $@ $(JSMATH_PRELINK) $(FDLIBM_LIBRARY)
+	$(AR) rc $(FDLIBM_LIBRARY) $@ $(JSMATH_PRELINK)
+#	$(LD) $(DASH_R) -o $@ $(JSMATH_PRELINK) $(FDLIBM_LIBRARY)
 endif
 	@$(RM) -f $(JSMATH_PRELINK)
 else
Index: js/src/jsfile.c
===================================================================
RCS file: /cvsroot/mozilla/js/src/jsfile.c,v
retrieving revision 3.36.4.4
diff -u -r3.36.4.4 jsfile.c
--- js/src/jsfile.c	1 Nov 2005 01:20:21 -0000	3.36.4.4
+++ js/src/jsfile.c	22 May 2006 00:40:11 -0000
@@ -1460,6 +1460,7 @@
     jsval		count, size;
     JSBool      fileInitiallyOpen=JS_FALSE;
 
+puts("file_copyto 1");
     SECURITY_CHECK(cx, NULL, "copyTo", file);   /* may need a second argument!*/
     JSFILE_CHECK_ONE_ARG("copyTo");
     JSFILE_CHECK_NATIVE("copyTo");
@@ -1467,6 +1468,7 @@
     fileInitiallyOpen = file->isOpen;
     JSFILE_CHECK_READ;
 
+puts("file_copyto 2");
     dest = JS_GetStringBytes(JS_ValueToString(cx, argv[0]));
 
     /* make sure we are not reading a file open for writing */
@@ -1476,6 +1478,7 @@
         goto out;
     }
 
+puts("file_copyto 3");
     if (file->handle==NULL){
         JS_ReportErrorNumber(cx, JSFile_GetErrorMessage, NULL,
             JSFILEMSG_OP_FAILED, "open", file->path);
@@ -1484,12 +1487,14 @@
 
     handle = PR_Open(dest, PR_WRONLY|PR_CREATE_FILE|PR_TRUNCATE, 0644);
 
+puts("file_copyto 4");
     if(!handle){
         JS_ReportErrorNumber(cx, JSFile_GetErrorMessage, NULL,
             JSFILEMSG_OP_FAILED, "open", dest);
         goto out;
     }
 
+puts("file_copyto 5");
     if ((size=js_size(cx, file))==JSVAL_VOID) {
         goto out;
     }
@@ -1505,6 +1510,7 @@
               JSFILEMSG_COPY_READ_ERROR, file->path);
         goto out;
     }
+puts("file_copyto 6");
 
     count = INT_TO_JSVAL(PR_Write(handle, buffer, JSVAL_TO_INT(size)));
 
@@ -1522,6 +1528,7 @@
 		if(!file_close(cx, obj, 0, NULL, rval)) goto out;
 	}
 
+puts("file_copyto 7");
     if(PR_Close(handle)!=PR_SUCCESS){
         JS_ReportErrorNumber(cx, JSFile_GetErrorMessage, NULL,
               JSFILEMSG_OP_FAILED, "close", dest);
@@ -1531,6 +1538,7 @@
     *rval = JSVAL_TRUE;
     return JS_TRUE;
 out:
+puts("file_copyto out");
     if(file->isOpen && !fileInitiallyOpen){
         if(PR_Close(file->handle)!=PR_SUCCESS){
             JS_ReportWarning(cx, "Can't close %s, proceeding", file->path);
Index: parser/htmlparser/src/COtherElements.h
===================================================================
RCS file: /cvsroot/mozilla/parser/htmlparser/src/Attic/COtherElements.h,v
retrieving revision 3.73
diff -u -r3.73 COtherElements.h
--- parser/htmlparser/src/COtherElements.h	1 Jun 2005 22:23:12 -0000	3.73
+++ parser/htmlparser/src/COtherElements.h	22 May 2006 00:40:43 -0000
@@ -419,13 +419,13 @@
 public:
 
   static CGroupMembers& GetGroup(void) {
-    static CGroupMembers theGroup={0};
+    static CGroupMembers theGroup; //={0};
     theGroup.mBits.mInlineEntity=1;
     return theGroup;
   }
 
   static CGroupMembers& GetContainedGroups(void) {
-    static CGroupMembers theGroup={0};
+    static CGroupMembers theGroup; //={0};
     static PRBool initialized=PR_FALSE;
     if(!initialized) {
       initialized=PR_TRUE;
Index: rdf/base/src/nsRDFParserUtils.cpp
===================================================================
RCS file: /cvsroot/mozilla/rdf/base/src/Attic/nsRDFParserUtils.cpp,v
retrieving revision 1.17
diff -u -r1.17 nsRDFParserUtils.cpp
--- rdf/base/src/nsRDFParserUtils.cpp	17 Apr 2004 18:29:58 -0000	1.17
+++ rdf/base/src/nsRDFParserUtils.cpp	22 May 2006 00:40:45 -0000
@@ -149,6 +149,7 @@
                 if (cp - cbuf > 5) {
                     continue;
                 }
+#undef atoi
                 PRInt32 ch = PRInt32( ::atoi(cbuf) );
                 if (ch > 65535) {
                     continue;
Index: security/coreconf/config.mk
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/config.mk,v
retrieving revision 1.17
diff -u -r1.17 config.mk
--- security/coreconf/config.mk	25 Apr 2004 15:02:17 -0000	1.17
+++ security/coreconf/config.mk	22 May 2006 00:40:45 -0000
@@ -63,7 +63,7 @@
 #######################################################################
 
 TARGET_OSES = FreeBSD BSD_OS NetBSD OpenUNIX OS2 QNX Darwin BeOS OpenBSD \
-              OpenVMS AIX
+              OpenVMS AIX riscos
 
 ifeq (,$(filter-out $(TARGET_OSES),$(OS_TARGET)))
 include $(CORE_DEPTH)/coreconf/$(OS_TARGET).mk
Index: security/coreconf/rules.mk
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/rules.mk,v
retrieving revision 1.62
diff -u -r1.62 rules.mk
--- security/coreconf/rules.mk	15 Apr 2005 00:36:28 -0000	1.62
+++ security/coreconf/rules.mk	22 May 2006 00:40:46 -0000
@@ -350,8 +350,12 @@
 ifdef XP_OS2_VACPP
 	$(MKSHLIB) $(DLLFLAGS) $(LDFLAGS) $(OBJS) $(SUB_SHLOBJS) $(LD_LIBS) $(EXTRA_LIBS) $(EXTRA_SHARED_LIBS)
 else
+ifeq ($(OS_TARGET),riscos)
+	$(MKSHLIB) $@ $(OBJS) $(SUB_SHLOBJS)
+else
 	$(MKSHLIB) -o $@ $(OBJS) $(SUB_SHLOBJS) $(LD_LIBS) $(EXTRA_LIBS) $(EXTRA_SHARED_LIBS)
 endif
+endif
 	chmod +x $@
 ifeq ($(OS_TARGET),Darwin)
 ifdef MAPFILE
Index: security/coreconf/nsinstall/nsinstall.c
===================================================================
RCS file: /cvsroot/mozilla/security/coreconf/nsinstall/nsinstall.c,v
retrieving revision 1.8
diff -u -r1.8 nsinstall.c
--- security/coreconf/nsinstall/nsinstall.c	25 Apr 2004 15:02:18 -0000	1.8
+++ security/coreconf/nsinstall/nsinstall.c	22 May 2006 00:40:47 -0000
@@ -58,7 +58,7 @@
 
 #define HAVE_LCHOWN
 
-#if defined(AIX) || defined(BSDI) || defined(HPUX) || defined(LINUX) || defined(SUNOS4) || defined(SCO) || defined(UNIXWARE) || defined(VMS) || defined(NTO) || defined(DARWIN) || defined(BEOS)
+#if defined(AIX) || defined(BSDI) || defined(HPUX) || defined(LINUX) || defined(SUNOS4) || defined(SCO) || defined(UNIXWARE) || defined(VMS) || defined(NTO) || defined(DARWIN) || defined(BEOS) || defined(__riscos__)
 #undef HAVE_LCHOWN
 #endif
 
@@ -68,9 +68,9 @@
 #undef HAVE_FCHMOD
 #endif
 
-#ifdef LINUX
+/*#if defined(LINUX)*/
 #include <getopt.h>
-#endif
+/*#endif*/
 
 #if defined(SCO) || defined(UNIXWARE) || defined(SNI) || defined(NCR) || defined(NEC)
 #if !defined(S_ISLNK) && defined(S_IFLNK)
Index: security/manager/ssl/src/nsKeygenHandler.cpp
===================================================================
RCS file: /cvsroot/mozilla/security/manager/ssl/src/nsKeygenHandler.cpp,v
retrieving revision 1.36
diff -u -r1.36 nsKeygenHandler.cpp
--- security/manager/ssl/src/nsKeygenHandler.cpp	20 Jul 2005 19:31:22 -0000	1.36
+++ security/manager/ssl/src/nsKeygenHandler.cpp	22 May 2006 00:40:48 -0000
@@ -95,6 +95,7 @@
     { 0, }
 };
 
+#if 0
 DERTemplate SECAlgorithmIDTemplate[] = {
     { DER_SEQUENCE,
 	  0, NULL, sizeof(SECAlgorithmID) },
@@ -112,6 +113,7 @@
     { SEC_ASN1_INTEGER, offsetof(PQGParams,base) },
     { 0, }
 };
+#endif
 
 
 static NS_DEFINE_IID(kIDOMHTMLSelectElementIID, NS_IDOMHTMLSELECTELEMENT_IID);
Index: security/nss/lib/freebl/unix_rand.c
===================================================================
RCS file: /cvsroot/mozilla/security/nss/lib/freebl/unix_rand.c,v
retrieving revision 1.16.14.1
diff -u -r1.16.14.1 unix_rand.c
--- security/nss/lib/freebl/unix_rand.c	10 Oct 2005 23:22:50 -0000	1.16.14.1
+++ security/nss/lib/freebl/unix_rand.c	22 May 2006 00:40:50 -0000
@@ -34,6 +34,8 @@
  *
  * ***** END LICENSE BLOCK ***** */
 
+#define __USE_BSD
+
 #include <stdio.h>
 #include <string.h>
 #include <signal.h>
@@ -81,7 +83,7 @@
 }
 
 #if defined(SCO) || defined(UNIXWARE) || defined(BSDI) || defined(FREEBSD) \
-    || defined(NETBSD) || defined(NTO) || defined(DARWIN) || defined(OPENBSD)
+    || defined(NETBSD) || defined(NTO) || defined(DARWIN) || defined(OPENBSD) || defined(__riscos__)
 #include <sys/times.h>
 
 #define getdtablesize() sysconf(_SC_OPEN_MAX)
Index: security/nss/lib/softoken/pcertdb.c
===================================================================
RCS file: /cvsroot/mozilla/security/nss/lib/softoken/pcertdb.c,v
retrieving revision 1.49
diff -u -r1.49 pcertdb.c
--- security/nss/lib/softoken/pcertdb.c	29 Mar 2005 18:21:18 -0000	1.49
+++ security/nss/lib/softoken/pcertdb.c	22 May 2006 00:40:54 -0000
@@ -3985,6 +3985,11 @@
 #define NO_CREATE	(O_RDWR | O_CREAT | O_TRUNC)
 #endif
 
+DB *dbopen(const char *file, int flags, int mode, DBTYPE type, const void *openinfo)
+{
+    return __db185_open_4004(file, flags, mode, type, openinfo);
+} 
+
 /*
  * open an old database that needs to be updated
  */
Index: toolkit/profile/src/nsToolkitProfileService.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/profile/src/nsToolkitProfileService.cpp,v
retrieving revision 1.12.2.1
diff -u -r1.12.2.1 nsToolkitProfileService.cpp
--- toolkit/profile/src/nsToolkitProfileService.cpp	19 Aug 2005 19:36:57 -0000	1.12.2.1
+++ toolkit/profile/src/nsToolkitProfileService.cpp	22 May 2006 00:41:01 -0000
@@ -305,11 +305,18 @@
 {
     nsresult rv;
 
+//    mDirectory = aDirectory;
+//    return 0;
     rv = mLock.Lock(aDirectory, aUnlocker);
 
     if (NS_SUCCEEDED(rv)) {
         mDirectory = aDirectory;
         mLocalDirectory = aLocalDirectory;
+    } else {
+      printf("profile lock failed\n");
+        mDirectory = aDirectory;
+        mLocalDirectory = aLocalDirectory;
+      return 0;
     }
 
     return rv;
Index: toolkit/xre/nsAppRunner.cpp
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsAppRunner.cpp,v
retrieving revision 1.113.2.4.2.4
diff -u -r1.113.2.4.2.4 nsAppRunner.cpp
--- toolkit/xre/nsAppRunner.cpp	6 Apr 2006 17:27:00 -0000	1.113.2.4.2.4
+++ toolkit/xre/nsAppRunner.cpp	22 May 2006 00:41:04 -0000
@@ -1207,9 +1207,20 @@
   if (_execv(exePath.get(), gRestartArgv) == -1)
     return NS_ERROR_FAILURE;
 #elif defined(XP_UNIX)
+  printf("LaunchChild: %s\n", exePath.get());
+
+  {
+    int arg = 0;
+
+    while (gRestartArgv[arg]) {
+      printf("arg: %s\n", gRestartArgv[arg++]);
+    }
+  }
+
   if (execv(exePath.get(), gRestartArgv) == -1)
     return NS_ERROR_FAILURE;
 #else
+
   PRProcess* process = PR_CreateProcess(exePath.get(), gRestartArgv,
                                         nsnull, nsnull);
   if (!process) return NS_ERROR_FAILURE;
@@ -1430,15 +1441,20 @@
   *aResult = nsnull;
   *aStartOffline = PR_FALSE;
 
+puts("profile1");
   arg = PR_GetEnv("XRE_START_OFFLINE");
   if ((arg && *arg) || CheckArg("offline"))
     *aStartOffline = PR_TRUE;
+puts("profile1a");
 
   arg = PR_GetEnv("XRE_PROFILE_PATH");
+puts("profile1b");
   if (arg && *arg) {
     nsCOMPtr<nsILocalFile> lf;
+puts("profile1bb");
     rv = NS_NewNativeLocalFile(nsDependentCString(arg), PR_TRUE,
                                getter_AddRefs(lf));
+puts("profile1bc");
     NS_ENSURE_SUCCESS(rv, rv);
 
     nsCOMPtr<nsILocalFile> localDir;
@@ -1459,6 +1475,7 @@
 
     return NS_LockProfilePath(lf, localDir, nsnull, aResult);
   }
+puts("profile2");
 
   if (CheckArg("migration"))
     gDoMigration = PR_TRUE;
@@ -1484,6 +1501,7 @@
     return ProfileLockedDialog(lf, lf, unlocker, aNative, aResult);
   }
 
+puts("profile3");
   nsCOMPtr<nsIToolkitProfileService> profileSvc;
   rv = NS_NewToolkitProfileService(getter_AddRefs(profileSvc));
   NS_ENSURE_SUCCESS(rv, rv);
@@ -1493,9 +1511,11 @@
     PR_fprintf(PR_STDERR, "Error: argument -createprofile requires a profile name\n");
     return NS_ERROR_FAILURE;
   }
+puts("profile4");
   if (ar) {
     nsCOMPtr<nsIToolkitProfile> profile;
 
+puts("profile4a");
     const char* delim = strchr(arg, ' ');
     if (delim) {
       nsCOMPtr<nsILocalFile> lf;
@@ -1514,6 +1534,7 @@
       rv = profileSvc->CreateProfile(nsnull, nsnull, nsDependentCString(arg),
                                      getter_AddRefs(profile));
     }
+puts("profile4b");
     if (NS_SUCCEEDED(rv)) {
       rv = NS_ERROR_ABORT;
       PR_fprintf(PR_STDERR, "Success: created profile '%s'\n", arg);
@@ -1522,6 +1543,7 @@
 
     // XXXben need to ensure prefs.js exists here so the tinderboxes will
     //        not go orange.
+puts("profile4c");
     nsCOMPtr<nsILocalFile> prefsJSFile;
     profile->GetRootDir(getter_AddRefs(prefsJSFile));
     prefsJSFile->AppendNative(NS_LITERAL_CSTRING("prefs.js"));
@@ -1533,22 +1555,28 @@
 
     return rv;
   }
+puts("profile5");
 
   PRUint32 count;
   rv = profileSvc->GetProfileCount(&count);
   NS_ENSURE_SUCCESS(rv, rv);
+puts("profile5a");
 
   if (gAppData->flags & NS_XRE_ENABLE_PROFILE_MIGRATOR) {
     arg = PR_GetEnv("XRE_IMPORT_PROFILES");
+puts("profile5b");
     if (!count && (!arg || !*arg)) {
+puts("profile5c");
       return ImportProfiles(profileSvc, aNative);
     }
   }
 
+puts("profile6");
   ar = CheckArg("p", &arg);
   if (ar == ARG_BAD) {
     return ShowProfileManager(profileSvc, aNative);
   }
+puts("profile7");
   if (ar) {
     nsCOMPtr<nsIToolkitProfile> profile;
     rv = profileSvc->GetProfileByName(nsDependentCString(arg),
@@ -1571,9 +1599,11 @@
                                  aNative, aResult);
     }
 
+puts("profile8");
     return ShowProfileManager(profileSvc, aNative);
   }
 
+puts("profile9");
   if (CheckArg("profilemanager")) {
     return ShowProfileManager(profileSvc, aNative);
   }
@@ -1581,6 +1611,7 @@
   if (!count) {
     gDoMigration = PR_TRUE;
 
+puts("profile10");
     // create a default profile
     nsCOMPtr<nsIToolkitProfile> profile;
     nsresult rv = profileSvc->CreateProfile(nsnull, // choose a default dir for us
@@ -1599,29 +1630,36 @@
   if (count > 1)
     profileSvc->GetStartWithLastProfile(&useDefault);
 
+puts("profile11");
   if (useDefault) {
     nsCOMPtr<nsIToolkitProfile> profile;
     // GetSelectedProfile will auto-select the only profile if there's just one
     profileSvc->GetSelectedProfile(getter_AddRefs(profile));
+puts("profile12");
     if (profile) {
       nsCOMPtr<nsIProfileUnlocker> unlocker;
       rv = profile->Lock(getter_AddRefs(unlocker), aResult);
+puts("profile13");
       if (NS_SUCCEEDED(rv))
         return NS_OK;
 
+puts("profile14");
       nsCOMPtr<nsILocalFile> profileDir;
       rv = profile->GetRootDir(getter_AddRefs(profileDir));
       NS_ENSURE_SUCCESS(rv, rv);
+puts("profile15");
 
       nsCOMPtr<nsILocalFile> profileLocalDir;
       rv = profile->GetRootDir(getter_AddRefs(profileLocalDir));
       NS_ENSURE_SUCCESS(rv, rv);
+puts("profile16");
 
       return ProfileLockedDialog(profileDir, profileLocalDir, unlocker,
                                  aNative, aResult);
     }
   }
 
+puts("profile7");
   return ShowProfileManager(profileSvc, aNative);
 }
 
@@ -1850,6 +1888,7 @@
 {
   nsresult rv;
   NS_TIMELINE_MARK("enter main");
+puts("main 1");
 
 #ifdef DEBUG
   if (PR_GetEnv("XRE_MAIN_BREAK"))
@@ -1874,9 +1913,11 @@
 #if defined(XP_UNIX) || defined(XP_BEOS)
   InstallUnixSignalHandlers(argv[0]);
 #endif
+puts("main 2");
 
   // Unbuffer stdout, needed for tinderbox tests.
   setbuf(stdout, 0);
+puts("main 3");
 
 #if defined(FREEBSD)
   // Disable all SIGFPE's on FreeBSD, as it has non-IEEE-conformant fp
@@ -1884,6 +1925,7 @@
   // the JS engine.  See bugzilla bug 9967 details.
   fpsetmask(0);
 #endif
+puts("main 4");
 
   gArgc = argc;
   gArgv = argv;
@@ -1913,6 +1955,7 @@
   gRestartArgv = (char**) malloc(sizeof(char*) * (argc + 1));
   if (!gRestartArgv) return 1;
 
+puts("main 5");
   int i;
   for (i = 0; i < argc; ++i) {
     gRestartArgv[i] = argv[i];
@@ -1945,6 +1988,7 @@
   gArgc = argc = NS_TraceMallocStartupArgs(argc, argv);
 #endif
 
+puts("main 6");
   nsXREDirProvider dirProvider;
   {
     rv = dirProvider.Initialize(gAppData->directory);
@@ -1952,6 +1996,7 @@
       return 1;
   }
 
+puts("main 7");
   // Check for -register, which registers chrome and then exits immediately.
   if (CheckArg("register")) {
     ScopedXPCOMStartup xpcom;
@@ -2007,6 +2052,7 @@
 // XXXtimeless fix me! How do we get a Display from here to nsAppShell.cpp ?
 // #endif
     
+puts("main 8");
   // Call the code to install our handler
 #ifdef MOZ_JPROF
   setupProfilingStuff();
@@ -2023,6 +2069,7 @@
   if (NS_FAILED(rv) || !canRun) {
     return 1;
   }
+puts("main 9");
 
   //----------------------------------------------------------------
   // We need to check if a previous installation occured and
@@ -2092,23 +2139,29 @@
                  gRestartArgc,
                  gRestartArgv);
 #endif
+puts("main 10");
 
   nsCOMPtr<nsIProfileLock> profileLock;
   PRBool startOffline = PR_FALSE;
 
+puts("main 10a");
   rv = SelectProfile(getter_AddRefs(profileLock), nativeApp, &startOffline);
   if (rv == NS_ERROR_LAUNCHED_CHILD_PROCESS ||
       rv == NS_ERROR_ABORT) return 0;
+  printf("%d %d\n", rv == NS_ERROR_LAUNCHED_CHILD_PROCESS, rv);
   if (NS_FAILED(rv)) return 1;
 
+puts("main 10b");
   nsCOMPtr<nsILocalFile> profD;
   rv = profileLock->GetDirectory(getter_AddRefs(profD));
   NS_ENSURE_SUCCESS(rv, 1);
 
+puts("main 10c");
   nsCOMPtr<nsILocalFile> profLD;
   rv = profileLock->GetLocalDirectory(getter_AddRefs(profLD));
   NS_ENSURE_SUCCESS(rv, 1);
 
+puts("main 10d");
   rv = dirProvider.SetProfile(profD, profLD);
   NS_ENSURE_SUCCESS(rv, 1);
 
@@ -2118,6 +2171,7 @@
 
   nsCAutoString version;
   BuildVersion(version);
+puts("main 11");
 
 #ifdef TARGET_OS_ABI
     NS_NAMED_LITERAL_CSTRING(osABI, TARGET_OS_ABI);
@@ -2171,6 +2225,7 @@
                  dirProvider.GetAppDir(), gAppData->directory);
   }
 
+puts("main 12");
   PRBool needsRestart = PR_FALSE;
   PRBool appInitiatedRestart = PR_FALSE;
 
@@ -2180,13 +2235,19 @@
   {
     // Start the real application
     ScopedXPCOMStartup xpcom;
+puts("main 12.1");
     rv = xpcom.Initialize();
-    NS_ENSURE_SUCCESS(rv, 1); 
+    NS_ENSURE_SUCCESS(rv, 1);
+puts("main 12.2");
     rv = xpcom.DoAutoreg();
+printf("rv: %x\n", rv);
     rv |= xpcom.InitEventQueue();
+printf("rv: %x\n", rv);
     rv |= xpcom.SetWindowCreator(nativeApp);
+printf("rv: %x\n", rv);
     NS_ENSURE_SUCCESS(rv, 1);
 
+puts("main 12a");
     {
       if (startOffline) {
         nsCOMPtr<nsIIOService> io (do_GetService("@mozilla.org/network/io-service;1"));
@@ -2194,6 +2255,7 @@
         io->SetOffline(PR_TRUE);
       }
 
+puts("main 12b");
       {
         NS_TIMELINE_ENTER("startupNotifier");
         nsCOMPtr<nsIObserver> startupNotifier
@@ -2229,6 +2291,7 @@
         }
       }
 
+puts("main 12c");
       // Profile Migration
       if (gAppData->flags & NS_XRE_ENABLE_PROFILE_MIGRATOR && gDoMigration) {
         gDoMigration = PR_FALSE;
@@ -2239,6 +2302,7 @@
       }
       dirProvider.DoStartup();
 
+puts("main 12c");
       nsCOMPtr<nsICommandLineRunner> cmdLine
         (do_CreateInstance("@mozilla.org/toolkit/command-line;1"));
       NS_ENSURE_TRUE(cmdLine, 1);
@@ -2251,6 +2315,7 @@
                          workingDir, nsICommandLine::STATE_INITIAL_LAUNCH);
       NS_ENSURE_SUCCESS(rv, 1);
 
+puts("main 13");
       /* Special-case services that need early access to the command
          line. */
       nsCOMPtr<nsIObserver> chromeObserver
@@ -2259,22 +2324,29 @@
         chromeObserver->Observe(cmdLine, "command-line-startup", nsnull);
       }
 
+puts("main 14");
       NS_TIMELINE_ENTER("appStartup->CreateHiddenWindow");
       rv = appStartup->CreateHiddenWindow();
       NS_TIMELINE_LEAVE("appStartup->CreateHiddenWindow");
       NS_ENSURE_SUCCESS(rv, 1);
 
+puts("main 15");
       // Extension Compatibility Checking and Startup
       if (gAppData->flags & NS_XRE_ENABLE_EXTENSION_MANAGER) {
         nsCOMPtr<nsIExtensionManager> em(do_GetService("@mozilla.org/extensions/manager;1"));
+puts("main 15a");
         NS_ENSURE_TRUE(em, 1);
 
+puts("main 15b");
         if (CheckArg("install-global-extension") || CheckArg("install-global-theme")) {
           // Do the required processing and then shut down.
+puts("main 15c");
           em->HandleCommandLineArgs(cmdLine);
+puts("main 15d");
           return 0;
         }
 
+puts("main 16");
         if (upgraded) {
           rv = em->CheckForMismatches(&needsRestart);
           if (NS_FAILED(rv)) {
@@ -2283,6 +2355,7 @@
           }
         }
 
+printf("upgrade 1: %d restart: %d\n", upgraded, needsRestart);
         if (!upgraded || !needsRestart)
           em->Start(cmdLine, &needsRestart);
       }
@@ -2297,6 +2370,9 @@
         needsRestart = upgraded = PR_FALSE;
       }
 
+printf("upgrade 2: %d restart: %d\n", upgraded, needsRestart);
+      needsRestart = upgraded = PR_FALSE;
+
       if (!upgraded && !needsRestart) {
         SaveStateForAppInitiatedRestart();
 
@@ -2409,6 +2485,7 @@
 
     profileLock->Unlock();
   }
+puts("main 14");
 
   // Restart the app after XPCOM has been shut down cleanly. 
   if (needsRestart) {
@@ -2437,8 +2514,7 @@
     }
 #endif
 
-    rv = LaunchChild(nativeApp, appInitiatedRestart);
-    return rv == NS_ERROR_LAUNCHED_CHILD_PROCESS ? 0 : 1;
+    return LaunchChild(nativeApp) == NS_ERROR_LAUNCHED_CHILD_PROCESS ? 0 : 1;
   }
 
   return NS_FAILED(rv) ? 1 : 0;
Index: toolkit/xre/nsAppRunner.h
===================================================================
RCS file: /cvsroot/mozilla/toolkit/xre/nsAppRunner.h,v
retrieving revision 1.17.2.2
diff -u -r1.17.2.2 nsAppRunner.h
--- toolkit/xre/nsAppRunner.h	14 Sep 2005 14:09:18 -0000	1.17.2.2
+++ toolkit/xre/nsAppRunner.h	22 May 2006 00:41:04 -0000
@@ -48,7 +48,7 @@
 #elif defined(CCHMAXPATH)
 #define MAXPATHLEN CCHMAXPATH
 #else
-#define MAXPATHLEN 1024
+#define MAXPATHLEN 256
 #endif
 #endif
 
Index: widget/src/xlib/nsWidget.cpp
===================================================================
RCS file: /cvsroot/mozilla/widget/src/xlib/nsWidget.cpp,v
retrieving revision 1.99.6.2
diff -u -r1.99.6.2 nsWidget.cpp
--- widget/src/xlib/nsWidget.cpp	20 Oct 2005 23:34:13 -0000	1.99.6.2
+++ widget/src/xlib/nsWidget.cpp	22 May 2006 00:41:10 -0000
@@ -412,25 +412,29 @@
     PRInt32 screenWidth = WidthOfScreen(mScreen);
     PRInt32 screenHeight = HeightOfScreen(mScreen);
 
+    printf("popup: %d %d\n", aX, aY);
+
     if (aX >= screenWidth)
       aX = screenWidth - mBounds.width;
     if (aY >= screenHeight)
       aY = screenHeight - mBounds.height;
 
-    aRect.x = aX;
-    aRect.y = aY;
-
-    if (mParentWidget) {
-      mParentWidget->WidgetToScreen(aRect, transRect);
-    } else if (mParentWindow) {
-      Window child;
-      XTranslateCoordinates(mDisplay, mParentWindow,
-                            XRootWindowOfScreen(mScreen),
-                            aX, aY, &transRect.x, &transRect.y,
-                            &child);
-    }
-    aX = transRect.x;
-    aY = transRect.y;
+//    aRect.x = aX;
+//    aRect.y = aY;
+//
+//    if (mParentWidget) {
+//      puts("parentWidget");
+//      mParentWidget->WidgetToScreen(aRect, transRect);
+//    } else if (mParentWindow) {
+//      Window child;
+//      puts("parentWindow");
+//      XTranslateCoordinates(mDisplay, mParentWindow,
+//                            XRootWindowOfScreen(mScreen),
+//                            aX, aY, &transRect.x, &transRect.y,
+//                            &child);
+//    }
+//    aX = transRect.x;
+//    aY = transRect.y;
   }
 
   mRequestedSize.x = aX;
Index: xpcom/build/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/xpcom/build/Makefile.in,v
retrieving revision 1.90.2.1
diff -u -r1.90.2.1 Makefile.in
--- xpcom/build/Makefile.in	19 Aug 2005 19:36:57 -0000	1.90.2.1
+++ xpcom/build/Makefile.in	22 May 2006 00:41:11 -0000
@@ -160,7 +160,7 @@
 		-DEXPORT_XPTC_API \
 		-DEXPORT_XPTI_API
 
-EXTRA_DSO_LDOPTS += $(NSPR_LIBS)
+#EXTRA_DSO_LDOPTS += $(NSPR_LIBS)
 
 ifdef GC_LEAK_DETECTOR
 DEFINES += -DGC_LEAK_DETECTOR
Index: xpcom/reflect/xptcall/src/md/unix/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/xpcom/reflect/xptcall/src/md/unix/Makefile.in,v
retrieving revision 1.80
diff -u -r1.80 Makefile.in
--- xpcom/reflect/xptcall/src/md/unix/Makefile.in	15 Jun 2005 08:32:21 -0000	1.80
+++ xpcom/reflect/xptcall/src/md/unix/Makefile.in	22 May 2006 00:41:12 -0000
@@ -161,6 +161,13 @@
 CPPSRCS		:= xptcinvoke_arm_netbsd.cpp xptcstubs_arm_netbsd.cpp
 endif
 endif
+#
+# RISC OS
+#
+ifeq ($(OS_ARCH),riscos)
+CPPSRCS		:= xptcinvoke_arm.cpp xptcstubs_arm_riscos.cpp
+CXXFLAGS += -O2
+endif
 
 ######################################################################
 # HPPA
Index: xpcom/reflect/xptcall/src/md/unix/xptcinvoke_arm.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/reflect/xptcall/src/md/unix/xptcinvoke_arm.cpp,v
retrieving revision 1.7.28.1
diff -u -r1.7.28.1 xptcinvoke_arm.cpp
--- xpcom/reflect/xptcall/src/md/unix/xptcinvoke_arm.cpp	13 Oct 2005 21:33:10 -0000	1.7.28.1
+++ xpcom/reflect/xptcall/src/md/unix/xptcinvoke_arm.cpp	22 May 2006 00:41:12 -0000
@@ -37,9 +37,10 @@
 
 /* Platform specific code to invoke XPCOM methods on native objects */
 
+#include <stdio.h>
 #include "xptcprivate.h"
 
-#if !defined(LINUX) || !defined(__arm__)
+#if (!defined(LINUX) && !defined(__riscos__)) || !defined(__arm__)
 #error "This code is for Linux ARM only. Check that it works on your system, too.\nBeware that this code is highly compiler dependent."
 #endif
 
@@ -144,6 +145,7 @@
 {
     PRUint32 result;
     struct my_params_struct my_params;
+
     my_params.that = that;
     my_params.Index = methodIndex;
     my_params.Count = paramCount;
@@ -177,43 +179,44 @@
  */
  
   __asm__ __volatile__(
-    "ldr	r1, [%1, #12]	\n\t"	/* prepare to call invoke_count_words	*/
-    "ldr	ip, [%1, #16]	\n\t"	/* r0=paramCount, r1=params		*/
-    "ldr	r0, [%1,  #8]	\n\t"
-    "mov	lr, pc		\n\t"	/* call it...				*/
+    "mov	r5, %1		\n\t"
+    "ldr	r1, [r5, #12]	\n\t"	/* prepare to call invoke_count_words	*/
+    "ldr	ip, [r5, #16]	\n\t"	/* r0=paramCount, r1=params		*/
+    "ldr	r4, [r5,  #8]	\n\t"
+    "mov	r0, r4		\n\t"
+    "mov	lr, pc		\n\t"	/* call it				*/
     "mov	pc, ip		\n\t"
-    "mov	r4, r0, lsl #2	\n\t"	/* This is the amount of bytes needed.	*/
-    "sub	sp, sp, r4	\n\t"	/* use stack space for the args...	*/
+    "sub	sp, sp, r4, lsl #2\n\t"	/* Use stack space for the args		*/
     "mov	r0, sp		\n\t"	/* prepare a pointer an the stack	*/
-    "ldr	r1, [%1,  #8]	\n\t"	/* =paramCount				*/
-    "ldr	r2, [%1, #12]	\n\t"	/* =params				*/
-    "ldr	ip, [%1, #20]	\n\t"	/* =invoke_copy_to_stack		*/
+    "ldr	r1, [r5,  #8]	\n\t"	/* =paramCount				*/
+    "ldr	r2, [r5, #12]	\n\t"	/* =params				*/
+    "ldr	ip, [r5, #20]	\n\t"	/* =invoke_copy_to_stack		*/
     "mov	lr, pc		\n\t"	/* copy args to the stack like the	*/
     "mov	pc, ip		\n\t"	/* compiler would.			*/
-    "ldr	r0, [%1]	\n\t"	/* =that				*/
+    "ldr	r0, [r5]	\n\t"	/* =that				*/
     "ldr	r1, [r0, #0]	\n\t"	/* get that->vtable offset		*/
-    "ldr	r2, [%1, #4]	\n\t"
-    "mov	r2, r2, lsl #2	\n\t"	/* a vtable_entry(x)=8 + (4 bytes * x)	*/
+    "ldr	r2, [r5, #4]	\n\t"
+ /*   "mov	r2, r2, lsl #2	\n\t"*/	/* a vtable_entry(x)=8 + (4 bytes * x)	*/
 #if defined(__GXX_ABI_VERSION) && __GXX_ABI_VERSION >= 100 /* G++ V3 ABI */
-    "ldr        ip, [r1, r2]    \n\t"   /* get method adress from vtable        */
+    "ldr        ip, [r1, r2, lsl #2]    \n\t"   /* get method adress from vtable        */
 #else /* non G++ V3 ABI */
-    "add	r2, r2, #8	\n\t"	/* with this compilers			*/
-    "ldr	ip, [r1, r2]	\n\t"	/* get method adress from vtable	*/
+    "add	r2, r2, #2	\n\t"	/* with these compilers			*/
+    "ldr	ip, [r1, r2, lsl #2]	\n\t"	/* get dmethod adress from vtable	*/
 #endif
-    "cmp	r4, #12		\n\t"	/* more than 3 arguments???		*/
+    "cmp	r4, #3		\n\t"	/* more than 3 arguments?		*/
     "ldmgtia	sp!, {r1, r2, r3}\n\t"	/* yes: load arguments for r1-r3	*/
-    "subgt	r4, r4, #12	\n\t"	/*      and correct the stack pointer	*/
+    "subgt	r4, r4, #3	\n\t"	/*      and correct the stack pointer	*/
     "ldmleia	sp, {r1, r2, r3}\n\t"	/* no:  load r1-r3 from stack		*/ 
-    "addle	sp, sp, r4	\n\t"	/*      and restore stack pointer	*/
+    "addle	sp, sp, r4, lsl #2\n\t"	/*      and restore stack pointer	*/
     "movle	r4, #0		\n\t"	/*	a mark for restoring sp		*/
-    "ldr	r0, [%1, #0]	\n\t"	/* get (self)				*/
-    "mov	lr, pc		\n\t"	/* call mathod				*/
+    "ldr	r0, [r5, #0]	\n\t"	/* get (self)				*/
+    "mov	lr, pc		\n\t"	/* call method				*/
     "mov	pc, ip		\n\t"
-    "add	sp, sp, r4	\n\t"	/* restore stack pointer		*/
-    "mov	%0, r0		\n\t"	/* the result...			*/
+    "add	sp, sp, r4, lsl #2\n\t"	/* restore stack pointer		*/
+    "mov	%0, r0		\n\t"	/* the result				*/
     : "=r" (result)
     : "r" (&my_params)
-    : "r0", "r1", "r2", "r3", "r4", "ip", "lr", "sp"
+    : "r0", "r1", "r2", "r3", "r4", "r5", "ip", "lr", "sp"
     );
     
   return result;
